File: experiments/problem1_nonperiodicity/analyze_period_two_signed_belief_derivative.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <limits>
#include <map>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

using u64 = std::uint64_t;
constexpr int MAXK = 28;
constexpr int CAP = 64;

u64 forward_generator(char g, u64 s) {
    const u64 t = s ^ ((s << 1) | (s << 2));
    if (g == 't') return t;
    if (g == 'u') return t ^ 1;
    if (g == 'p') return t ^ 1 ^ ((s & 1) == 0 ? 2 : 0);
    throw std::runtime_error("unknown generator");
}

std::vector<u64> next_level(const std::vector<u64>& current) {
    std::vector<u64> output;
    output.reserve(current.size() * 3);
    for (u64 state : current) {
        output.push_back(forward_generator('t', state));
        output.push_back(forward_generator('u', state));
        output.push_back(forward_generator('p', state));
    }
    std::sort(output.begin(), output.end());
    output.erase(std::unique(output.begin(), output.end()), output.end());
    return output;
}

bool contains(const std::vector<u64>& values, u64 target) {
    return std::binary_search(values.begin(), values.end(), target);
}

unsigned fiber_mask(
    const std::vector<std::vector<u64>>& levels, int complexity, u64 quotient
) {
    unsigned mask = 0;
    for (int digit = 0; digit < 4; ++digit) {
        if (contains(levels[complexity + 1], 4 * quotient + unsigned(digit))) {
            mask |= 1u << digit;
        }
    }
    return mask;
}

u64 mask_sequence(
    const std::vector<std::vector<u64>>& levels,
    int complexity,
    u64 state,
    int depth
) {
    u64 result = 0;
    for (int step = 0; step < depth; ++step) {
        state >>= 2;
        result |= u64(fiber_mask(levels, complexity - 1 - step, state))
            << (4 * step);
    }
    return result;
}

bool dominates(u64 current, u64 shadow, int depth) {
    for (int step = 0; step < depth; ++step) {
        const unsigned a = (current >> (4 * step)) & 15;
        const unsigned b = (shadow >> (4 * step)) & 15;
        if (a & ~b) return false;
    }
    return true;
}

bool synchronized_or_full(u64 current, u64 shadow, int depth) {
    for (int step = 0; step < depth; ++step) {
        const unsigned a = (current >> (4 * step)) & 15;
        const unsigned b = (shadow >> (4 * step)) & 15;
        if (b != 15 && b != a) return false;
    }
    return true;
}

int defect_count(u64 shadow, int depth) {
    int result = 0;
    for (int step = 0; step < depth; ++step) {
        result += ((shadow >> (4 * step)) & 15) != 15;
    }
    return result;
}

int local_signed_factor(unsigned current, unsigned shadow) {
    if (current & ~shadow) return 0;
    return shadow == 15 ? 1 : -1;
}

int local_branching_derivative(unsigned current, unsigned shadow) {
    if (current & ~shadow) return 0;
    if (current == 15) return shadow == 15 ? 1 : 0;
    if (current == 3 || current == 11) {
        return 2 * int((shadow >> 2) & 1) - 1;
    }
    if (current == 12) {
        return 2 * int(shadow & 1) - 1;
    }
    throw std::runtime_error("zero current mask entered relevant derivative");
}

std::string forced_zero_schedule(u64 state) {
    std::string word;
    for (int step = 0; step < CAP; ++step) {
        const u64 residue = state & 15;
        char branch;
        if (residue == 7) branch = 'u';
        else if (residue == 11) branch = 't';
        else return word;
        state = forward_generator(
            branch, forward_generator('p', (state - 3) >> 2)
        );
        word.push_back(branch);
    }
    throw std::runtime_error("schedule cap reached");
}

bool admissible(const std::string& word) {
    return word.find("uu") == std::string::npos
        && word.find("ttttt") == std::string::npos
        && word.find("ututtu") == std::string::npos;
}

struct Occurrence {
    int complexity;
    int cut;
    int depth;
    u64 current;
    u64 residue;
    u64 current_sequence;
    u64 same_cylinder = 0;
    u64 dominant = 0;
    u64 synchronized = 0;
    u64 budget_three = 0;
    int minimum_defect = 999;
    long long signed_mass = 0;
};

std::vector<Occurrence> process_level(
    const std::vector<std::vector<u64>>& levels, int complexity
) {
    std::vector<Occurrence> rows;
    int maximum_depth = 0;
    std::vector<std::unordered_map<u64, std::vector<int>>> lookup(24);

    for (u64 current : levels[complexity]) {
        if ((current & 3) != 3) continue;
        const std::string schedule = forced_zero_schedule(current);
        for (int cut = 0; cut <= static_cast<int>(schedule.size()); ++cut) {
            if (schedule.compare(cut, 6, "ututut") != 0) continue;
            if (!admissible(schedule.substr(0, cut) + "utututu")) continue;
            const int depth = cut + 1;
            const u64 modulus_mask = (u64{1} << (2 * depth)) - 1;
            const int index = rows.size();
            rows.push_back(Occurrence{
                complexity,
                cut,
                depth,
                current,
                current & modulus_mask,
                mask_sequence(levels, complexity, current, depth),
            });
            lookup[depth][current & modulus_mask].push_back(index);
            maximum_depth = std::max(maximum_depth, depth);
        }
    }

    for (u64 shadow : levels[complexity - 1]) {
        u64 state = shadow;
        u64 sequence = 0;
        int defects = 0;
        for (int step = 0; step < maximum_depth; ++step) {
            state >>= 2;
            const unsigned mask = fiber_mask(
                levels, complexity - 2 - step, state
            );
            sequence |= u64(mask) << (4 * step);
            defects += mask != 15;
            const int depth = step + 1;
            const u64 residue = shadow & ((u64{1} << (2 * depth)) - 1);
            const auto found = lookup[depth].find(residue);
            if (found == lookup[depth].end()) continue;
            for (int index : found->second) {
                auto& row = rows[index];
                ++row.same_cylinder;
                if (!dominates(row.current_sequence, sequence, depth)) continue;
                ++row.dominant;
                if (!synchronized_or_full(
                        row.current_sequence, sequence, depth
                    )) {
                    continue;
                }
                ++row.synchronized;
                row.minimum_defect = std::min(row.minimum_defect, defects);
                if (defects <= 3) ++row.budget_three;
                row.signed_mass += (defects & 1) ? -1 : 1;
            }
        }
    }
    return rows;
}

long long direct_signed_mass(
    const std::vector<std::vector<u64>>& levels,
    int complexity,
    u64 current,
    int depth
) {
    const u64 modulus_mask = (u64{1} << (2 * depth)) - 1;
    const u64 residue = current & modulus_mask;
    const u64 current_sequence = mask_sequence(levels, complexity, current, depth);
    long long result = 0;
    for (u64 shadow : levels[complexity - 1]) {
        if ((shadow & modulus_mask) != residue) continue;
        const u64 shadow_sequence = mask_sequence(
            levels, complexity - 1, shadow, depth
        );
        if (!dominates(current_sequence, shadow_sequence, depth)) continue;
        result += (defect_count(shadow_sequence, depth) & 1) ? -1 : 1;
    }
    return result;
}

int main() {
    for (unsigned current : {3u, 11u, 12u, 15u}) {
        for (unsigned shadow : {0u, 3u, 11u, 12u, 15u}) {
            if (local_signed_factor(current, shadow)
                != local_branching_derivative(current, shadow)) {
                throw std::runtime_error("local derivative identity failed");
            }
        }
    }

    std::vector<std::vector<u64>> levels(MAXK + 1);
    levels[1] = {1};
    u64 outputs = 1;
    for (int complexity = 2; complexity <= MAXK; ++complexity) {
        levels[complexity] = next_level(levels[complexity - 1]);
        outputs += levels[complexity].size();
    }

    u64 occurrences_27 = 0;
    u64 occurrences_28 = 0;
    u64 dominant_failures = 0;
    u64 synchronized_failures = 0;
    u64 signed_zeros = 0;
    u64 negative_signed = 0;
    u64 majority_failures = 0;
    std::map<int, u64> minimum_defects;
    long long minimum_absolute_signed = std::numeric_limits<long long>::max();
    Occurrence signed_witness{};
    u64 minimum_budget_count = std::numeric_limits<u64>::max();
    Occurrence budget_witness{};

    for (int complexity = 2; complexity <= MAXK; ++complexity) {
        auto rows = process_level(levels, complexity);
        if (complexity <= 27) occurrences_27 += rows.size();
        else occurrences_28 += rows.size();
        for (const auto& row : rows) {
            dominant_failures += row.dominant == 0;
            synchronized_failures += row.dominant - row.synchronized;
            signed_zeros += row.signed_mass == 0;
            negative_signed += row.signed_mass < 0;
            majority_failures += 2 * row.dominant < row.same_cylinder;
            ++minimum_defects[row.minimum_defect];
            const long long absolute = std::llabs(row.signed_mass);
            if (absolute < minimum_absolute_signed) {
                minimum_absolute_signed = absolute;
                signed_witness = row;
            }
            if (row.budget_three < minimum_budget_count) {
                minimum_budget_count = row.budget_three;
                budget_witness = row;
            }
        }
    }

    const long long cancellation = direct_signed_mass(levels, 5, 0x198, 1);

    std::cout << "maximum_complexity=" << MAXK << '\n';
    std::cout << "phase_u_outputs=" << outputs << '\n';
    std::cout << "gap_222_occurrences_through_27=" << occurrences_27 << '\n';
    std::cout << "gap_222_occurrences_at_28=" << occurrences_28 << '\n';
    std::cout << "gap_222_occurrences_total="
              << occurrences_27 + occurrences_28 << '\n';
    std::cout << "dominant_failures=" << dominant_failures << '\n';
    std::cout << "synchronized_failures=" << synchronized_failures << '\n';
    std::cout << "signed_zero_cylinders=" << signed_zeros << '\n';
    std::cout << "negative_signed_cylinders=" << negative_signed << '\n';
    std::cout << "majority_failures=" << majority_failures << '\n';
    for (const auto& [defects, count] : minimum_defects) {
        std::cout << "minimum_defect_" << defects << '=' << count << '\n';
    }
    std::cout << "minimum_absolute_signed_mass="
              << minimum_absolute_signed << '\n';
    std::cout << "minimum_budget_three_count="
              << minimum_budget_count << '\n';
    std::cout << std::hex;
    std::cout << "signed_witness_state=0x" << signed_witness.current << '\n';
    std::cout << "budget_witness_state=0x" << budget_witness.current << '\n';
    std::cout << std::dec;
    std::cout << "signed_witness_complexity=" << signed_witness.complexity
              << " cut=" << signed_witness.cut
              << " depth=" << signed_witness.depth
              << " mass=" << signed_witness.signed_mass << '\n';
    std::cout << "budget_witness_complexity=" << budget_witness.complexity
              << " cut=" << budget_witness.cut
              << " depth=" << budget_witness.depth
              << " budget_count=" << budget_witness.budget_three << '\n';
    std::cout << "nongap_cancellation_complexity_5_state_0x198_depth_1="
              << cancellation << '\n';

    if (occurrences_27 != 2989 || occurrences_28 != 2173
        || dominant_failures != 0 || synchronized_failures != 0
        || signed_zeros != 0 || minimum_defects[0] != 5159
        || minimum_defects[3] != 3 || minimum_absolute_signed != 1
        || minimum_budget_count != 1 || cancellation != 0) {
        throw std::runtime_error("signed-belief totals changed");
    }
}

File: experiments/problem1_nonperiodicity/analyze_period_two_signed_belief_derivative.py
#!/usr/bin/env python3
"""Signed defect mass and branching derivatives for concrete shadow beliefs."""
from __future__ import annotations

import argparse
import hashlib
import json
from collections import Counter
from typing import Any

DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 18
SCHEDULE_CAP = 64
MASK_ALPHABET = (0b0000, 0b0011, 0b1011, 0b1100, 0b1111)
RELEVANT_CURRENT_MASKS = (0b0011, 0b1011, 0b1100, 0b1111)


class SignedBeliefLimitError(RuntimeError):
    pass


def forward_generator(name: str, state: int) -> int:
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError(name)


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(name, state) for name in "tup"}))


def build_levels(maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {1}]
    for _ in range(2, maximum_complexity + 1):
        levels.append(
            {child for state in levels[-1] for child in frontier_children(state)}
        )
    return levels


def fiber_mask(levels: list[set[int]], complexity: int, quotient: int) -> int:
    mask = sum(
        1 << digit
        for digit in range(4)
        if 4 * quotient + digit in levels[complexity + 1]
    )
    if mask not in MASK_ALPHABET:
        raise AssertionError("fiber escaped the five-mask alphabet")
    return mask


def mask_sequence(
    levels: list[set[int]], complexity: int, state: int, depth: int
) -> tuple[int, ...]:
    result: list[int] = []
    for step in range(depth):
        state >>= 2
        result.append(fiber_mask(levels, complexity - 1 - step, state))
    return tuple(result)


def forced_zero_schedule(state: int, cap: int = SCHEDULE_CAP) -> str:
    word: list[str] = []
    for _ in range(cap):
        residue = state & 15
        if residue == 7:
            branch = "u"
        elif residue == 11:
            branch = "t"
        else:
            return "".join(word)
        state = forward_generator(
            branch, forward_generator("p", (state - 3) >> 2)
        )
        word.append(branch)
    raise SignedBeliefLimitError("forced schedule reached safety cap")


def admissible(word: str) -> bool:
    return all(factor not in word for factor in ("uu", "ttttt", "ututtu"))


def dominates(current: tuple[int, ...], shadow: tuple[int, ...]) -> bool:
    return all(
        not (current_mask & ~shadow_mask)
        for current_mask, shadow_mask in zip(current, shadow)
    )


def synchronized_or_full(
    current: tuple[int, ...], shadow: tuple[int, ...]
) -> bool:
    return all(
        shadow_mask == 0b1111 or shadow_mask == current_mask
        for current_mask, shadow_mask in zip(current, shadow)
    )


def defect_count(shadow: tuple[int, ...]) -> int:
    return sum(mask != 0b1111 for mask in shadow)


def local_signed_factor(current_mask: int, shadow_mask: int) -> int:
    """Return the local contribution to P(-1), including dominance rejection."""
    if current_mask & ~shadow_mask:
        return 0
    return 1 if shadow_mask == 0b1111 else -1


def local_branching_derivative(current_mask: int, shadow_mask: int) -> int:
    """Express the signed factor through one missing sibling indicator."""
    if current_mask & ~shadow_mask:
        return 0
    if current_mask == 0b1111:
        return int(shadow_mask == 0b1111)
    if current_mask in (0b0011, 0b1011):
        return 2 * ((shadow_mask >> 2) & 1) - 1
    if current_mask == 0b1100:
        return 2 * (shadow_mask & 1) - 1
    raise ValueError("branching derivative excludes current mask 0000")


def branching_path_weight(
    current: tuple[int, ...], shadow: tuple[int, ...]
) -> int:
    result = 1
    for current_mask, shadow_mask in zip(current, shadow):
        result *= local_branching_derivative(current_mask, shadow_mask)
    return result


def concrete_belief(
    levels: list[set[int]], complexity: int, current: int, depth: int
) -> tuple[tuple[int, ...], list[tuple[int, int, tuple[int, ...]]], int]:
    modulus = 4**depth
    residue = current % modulus
    current_masks = mask_sequence(levels, complexity, current, depth)
    rows: list[tuple[int, int, tuple[int, ...]]] = []
    same_cylinder = 0
    for shadow in levels[complexity - 1]:
        if shadow % modulus != residue:
            continue
        same_cylinder += 1
        shadow_masks = mask_sequence(levels, complexity - 1, shadow, depth)
        if not dominates(current_masks, shadow_masks):
            continue
        rows.append((shadow, defect_count(shadow_masks), shadow_masks))
    rows.sort()
    return current_masks, rows, same_cylinder


def signed_mass(rows: list[tuple[int, int, tuple[int, ...]]]) -> int:
    return sum(-1 if defects & 1 else 1 for _, defects, _ in rows)


def gap_222_occurrences(
    levels: list[set[int]], maximum_complexity: int
) -> list[dict[str, Any]]:
    result: list[dict[str, Any]] = []
    for complexity in range(2, maximum_complexity + 1):
        for current in sorted(levels[complexity]):
            if current & 3 != 3:
                continue
            schedule = forced_zero_schedule(current)
            for cut in range(len(schedule) + 1):
                if schedule[cut : cut + 6] != "ututut":
                    continue
                if not admissible(schedule[:cut] + "utututu"):
                    continue
                depth = cut + 1
                current_masks, belief, same_cylinder = concrete_belief(
                    levels, complexity, current, depth
                )
                if any(mask not in RELEVANT_CURRENT_MASKS for mask in current_masks):
                    raise AssertionError("relevant path acquired current mask 0000")
                synchronized_failures = sum(
                    not synchronized_or_full(current_masks, shadow_masks)
                    for _, _, shadow_masks in belief
                )
                derivative_disagreements = sum(
                    branching_path_weight(current_masks, shadow_masks)
                    != (-1 if defects & 1 else 1)
                    for _, defects, shadow_masks in belief
                )
                histogram = Counter(defects for _, defects, _ in belief)
                mass = signed_mass(belief)
                result.append(
                    {
                        "complexity": complexity,
                        "state_hex": hex(current),
                        "cut": cut,
                        "depth": depth,
                        "same_cylinder": same_cylinder,
                        "dominant_shadows": len(belief),
                        "budget_three_shadows": sum(
                            defects <= 3 for _, defects, _ in belief
                        ),
                        "minimum_defect": min(
                            (defects for _, defects, _ in belief), default=None
                        ),
                        "signed_mass": mass,
                        "current_masks": [f"{mask:04b}" for mask in current_masks],
                        "defect_histogram": {
                            str(key): value for key, value in sorted(histogram.items())
                        },
                        "synchronized_failures": synchronized_failures,
                        "derivative_disagreements": derivative_disagreements,
                    }
                )
    return result


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
) -> dict[str, Any]:
    if not 5 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise SignedBeliefLimitError("maximum complexity outside controlled range")

    for current_mask in RELEVANT_CURRENT_MASKS:
        for shadow_mask in MASK_ALPHABET:
            if local_signed_factor(current_mask, shadow_mask) != (
                local_branching_derivative(current_mask, shadow_mask)
            ):
                raise AssertionError("local branching derivative identity failed")

    levels = build_levels(maximum_complexity)
    rows = gap_222_occurrences(levels, maximum_complexity)
    if not rows:
        raise AssertionError("controlled campaign found no gap-222 occurrences")

    cancellation_masks, cancellation_belief, _ = concrete_belief(
        levels, 5, 0x198, 1
    )
    cancellation_histogram = Counter(
        defects for _, defects, _ in cancellation_belief
    )
    cancellation_mass = signed_mass(cancellation_belief)

    minimum_defects = Counter(row["minimum_defect"] for row in rows)
    payload: dict[str, Any] = {
        "status": "exact-signed-belief-branching-derivative-and-finite-nonvanishing",
        "phase": "u",
        "maximum_complexity": maximum_complexity,
        "outputs_built": sum(
            len(levels[complexity])
            for complexity in range(1, maximum_complexity + 1)
        ),
        "gap_222_occurrences": len(rows),
        "dominant_failures": sum(row["dominant_shadows"] == 0 for row in rows),
        "synchronized_failures": sum(
            row["synchronized_failures"] for row in rows
        ),
        "derivative_disagreements": sum(
            row["derivative_disagreements"] for row in rows
        ),
        "signed_zero_cylinders": sum(row["signed_mass"] == 0 for row in rows),
        "negative_signed_cylinders": sum(row["signed_mass"] < 0 for row in rows),
        "majority_failures": sum(
            2 * row["dominant_shadows"] < row["same_cylinder"] for row in rows
        ),
        "minimum_absolute_signed_mass": min(
            abs(row["signed_mass"]) for row in rows
        ),
        "minimum_budget_three_count": min(
            row["budget_three_shadows"] for row in rows
        ),
        "minimum_defect_histogram": {
            str(key): value for key, value in sorted(minimum_defects.items())
        },
        "theorem": {
            "defect_polynomial": (
                "For a concrete dominant shadow belief, P(z) is the sum of "
                "z raised to the number of non-full shadow fibers."
            ),
            "signed_mass": (
                "P(-1) is the sum over same-cylinder shadow endpoints of the "
                "product of local signed dominance factors. Nonzero P(-1) "
                "therefore certifies a nonempty concrete belief."
            ),
            "branching_derivative": (
                "On every relevant nonzero current mask, the local signed factor "
                "is exactly a Rademacher difference determined by one missing "
                "sibling child: digit 2 for masks 0011/1011 and digit 0 for "
                "mask 1100."
            ),
        },
        "nongap_cancellation": {
            "complexity": 5,
            "state_hex": "0x198",
            "depth": 1,
            "current_masks": [f"{mask:04b}" for mask in cancellation_masks],
            "defect_histogram": {
                str(key): value
                for key, value in sorted(cancellation_histogram.items())
            },
            "signed_mass": cancellation_mass,
            "conclusion": (
                "Signed nonvanishing is not universal; a proof must use the "
                "gap-return schedule rather than only the five-mask alphabet."
            ),
        },
        "examples": sorted(
            rows,
            key=lambda row: (
                abs(row["signed_mass"]),
                row["complexity"],
                row["state_hex"],
            ),
        )[:6],
        "scientific_boundary": (
            "The polynomial identity, branching-derivative identity, and "
            "nonzero-implies-nonempty implication are exact. Nonvanishing and "
            "the defect bounds are finite observations through the configured "
            "complexity and do not prove the all-depth adjacent-shadow inclusion."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity", type=int, default=DEFAULT_MAXIMUM_COMPLEXITY
    )
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_complexity), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
